pages/apps/address-book.tsx

```diff
@@ -137,8 +139,170 @@ function AddressBookPage(): ReactElement {
 	);
 }
 
+function AddressBookActions(): ReactElement {
+	const {addEntry, listEntries} = useAddressBook();
+
+	const handleFileUpload = (e: ChangeEvent<HTMLInputElement>): void => {
+		if (!e.target.files) {
+			return;
+		}
+		const [file] = e.target.files as unknown as Blob[];
+		const reader = new FileReader();
+		reader.onload = event => {
+			if (!event?.target?.result) {
+				return;
+			}
+			const {result} = event.target;
+			const parsedCSV = Papa.parse(result, {header: true});
+
+			// If we are working with a safe file, we should get 3 columns.
+			const isProbablySafeFile = parsedCSV.meta.fields.length === 3;
+			if (isProbablySafeFile) {
+				const [addressLike, name, chainID] = parsedCSV.meta.fields;
+				const records = parsedCSV.data.map((item: unknown[]) => {
+					return {
+						address: item[addressLike] as TAddress,
+						label: item[name] as string,
+						chains: [item[chainID]] as number[]
+					};
+				});
+
+				// On theses records, we might have the same address multiple times.
+				// If that's the case, we want to merge the chains and only keep one entry.
+				const mergedRecords = records.reduce((acc: TAddressBookEntry[], cur: TAddressBookEntry) => {
+					const existingRecord = acc.find(item => item.address === cur.address);
+					if (existingRecord) {
+						existingRecord.chains = [...existingRecord.chains, ...cur.chains];
+						return acc;
+					}
+					return [...acc, cur];
+				}, []);
+
+				// The name should always be unique. We need to check if we have duplicates.
+				// If that's the case, we need to add a slice of the address to the name, but
+				// we still keep both entries.
+				const uniqueRecords = mergedRecords.reduce((acc: TAddressBookEntry[], cur: TAddressBookEntry) => {
+					const existingRecord = acc.find(item => item.label === cur.label);
+					if (existingRecord) {
+						const existingSlice = toAddress(existingRecord.address).slice(0, 6);
+						existingRecord.label = `${existingRecord.label} (${existingSlice})`;
+
+						const curSlice = toAddress(cur.address).slice(0, 6);
+						cur.label = `${cur.label} (${curSlice})`;
+						return [...acc, cur];
+					}
+					return [...acc, cur];
+				}, []);
+				console.log(uniqueRecords);
+
+				for (const record of uniqueRecords) {
+					addEntry(record);
+				}
+			}
+
+			const isProbablySmolFile = parsedCSV.meta.fields.length === 4;
+			if (isProbablySmolFile) {
+				const [addressLike, name, chains, isFavorite] = parsedCSV.meta.fields;
+				const records = parsedCSV.data.map((item: unknown[]) => {
+					const chainIDs = ((item[chains] as string) || '').split(',').map(chain => Number(chain));
```

Comment:
Feels like chains handling is the only difference with `isProbablySafeFile` flow. Probably it is possible to move  `mergedRecords` and `uniqueRecords` into helpers that would work for both flows? just not to repeat the code